contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,407 @@
+        uint128 tokensOwed1;
+        uint256 blockNumber;
+        address liquidityToken;
+    }
+
+    mapping(PoolId => PoolInfo) public poolInfo;
+
+    constructor(IPoolManager _poolManager) BaseHook(_poolManager) {}
+
+    modifier ensure(uint256 deadline) {
+        require(deadline >= block.timestamp, "Expired");
+        _;
+    }
+
+    function getHooksCalls() public pure override returns (Hooks.Calls memory) {
+        return Hooks.Calls({
+            beforeInitialize: true,
+            afterInitialize: false,
+            beforeModifyPosition: true,
+            afterModifyPosition: false,
+            beforeSwap: true,
+            afterSwap: false,
+            beforeDonate: false,
+            afterDonate: false
+        });
+    }
+
+    function beforeInitialize(address, PoolKey calldata key, uint160) external override returns (bytes4) {
+        require(key.tickSpacing == 60, "Tick spacing must be default");
+
+        // deploy erc20 contract
+
+        // TODO: name, symbol for the ERC20 contract
+        bytes memory bytecode = type(UniswapV4ERC20).creationCode;
+        bytes32 salt = keccak256(abi.encodePacked(key.toId()));
+
+        address poolToken;
+        assembly {
+            poolToken := create2(0, add(bytecode, 0x20), mload(bytecode), salt)
+        }
+
+        PoolInfo memory info = PoolInfo({
+            liquidity: 0,
+            feeGrowthInside0LastX128: 0,
+            feeGrowthInside1LastX128: 0,
+            tokensOwed0: 0,
+            tokensOwed1: 0,
+            blockNumber: block.number,
+            liquidityToken: poolToken
+        });
+
+        poolInfo[key.toId()] = info;
+
+        return FullRange.beforeInitialize.selector;
+    }
+
+    function beforeModifyPosition(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.ModifyPositionParams calldata params
+    ) external override returns (bytes4) {
+        // check msg.sender
+        require(sender == address(this), "sender must be hook");
+        _rebalance(key);
+
+        return FullRange.beforeModifyPosition.selector;
+    }
+
+    function beforeSwap(address, PoolKey calldata key, IPoolManager.SwapParams calldata)
+        external
+        override
+        returns (bytes4)
+    {
+        // TODO: maybe don't sload the entire struct
+        PoolInfo storage position = poolInfo[key.toId()];
+        _rebalance(key);
+        return IHooks.beforeSwap.selector;
+    }
+
+    function balanceOf(Currency currency, address user) internal view returns (uint256) {
+        if (currency.isNative()) {
+            return user.balance;
+        } else {
+            return IERC20Minimal(Currency.unwrap(currency)).balanceOf(user);
+        }
+    }
+
+    function modifyPosition(PoolKey memory key, IPoolManager.ModifyPositionParams memory params)
+        internal
+        returns (BalanceDelta delta)
+    {
+        // msg.sender is the test contract (aka whoever called addLiquidity/removeLiquidity)
```

Comment:
nit: haha in prod it's not the test contract, I think this is self explanatory and comment not needed